### PR TITLE
GH-20484: [Swift] Initial Arrow implementation

### LINKED ISSUE #20484: [Swift] Swift implementation of Arrow
Initial check-in for a swift implementation of Arrow. Based on my understanding of the spec and looking through the C++ and C# current implementations.  

**Reporter**: [Alva Bandy](https://issues.apache.org/jira/browse/ARROW-18234) / @abandy
**Assignee**: [Alva Bandy](https://issues.apache.org/jira/browse/ARROW-18234) / @abandy
#### PRs and other links:
- [GitHub Pull Request #14561](https://github.com/apache/arrow/pull/14561)

<sub>**Note**: *This issue was originally created as [ARROW-18234](https://issues.apache.org/jira/browse/ARROW-18234). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### PR DESCRIPTION
**Pull request by abandy**
Initial check-in for a swift implementation of Arrow.  Based on my understanding of the spec and looking through the C++ and C# current implementations.
* Closes: #20484

### CODE REVIEW

**Comment by abandy** [comment]
I do not believe the component is correct but was unable to find Swift or a suitable alternative.

**Comment by kou** [comment]
Great!
Can we add a basic CI job for this implementation?

**Comment by abandy** [comment]
Great idea, will do!

**Comment by abandy** [comment]
I have been working on the swift arrow IPC impl to add to this PR.  Are there any test files in the arrow repo that I can use to test an IPC reader/writer?

**Comment by kou** [comment]
There are Apache Arrow files in https://github.com/apache/arrow-testing but most files are not for implementation testing...

We have a integration test framework: https://github.com/apache/arrow/tree/master/dev/archery/archery/integration
It compares outputs between other language implementations. Finally, this Swift implementation should join it but it may not be suitable for the first step...

> I have been working on the swift arrow IPC impl to add to this PR. Are there any test files in the arrow repo that I can use to test an IPC reader/writer?

How about split the implementation to another pull request? I don't want to review a large pull request as much as possible...

**Comment by abandy** [comment]
Awesome, will do, thank you!

File: .github/workflows/swift.yml
Comment (kou) [inline .github/workflows/swift.yml:62]: It seems that `matrix.ubuntu` isn't defined.

```suggestion
    env:
      UBUNTU: ${{ matrix.ubuntu || 20.04 }}
```

File: docker-compose.yml
Comment (kou) [inline docker-compose.yml:1427]: Could you revert this?

File: ci/docker/linux-apt-swift.dockerfile
Comment (kou) [inline ci/docker/linux-apt-swift.dockerfile:44]:
```suggestion
    ubuntu_version=$(. /etc/os-release && echo ${VERSION_ID}) && \
    ubuntu_version_no-dot=$(echo ${ubuntu_version} | sed -e 's/\.//g')
    wget https://download.swift.org/swift-5.7.3-release/ubuntu${ubuntu_version_no_dot}/swift-5.7.3-RELEASE/swift-5.7.3-RELEASE-ubuntu${ubuntu_version}.tar.gz && \
    tar -xvzf swift-5.7.3-RELEASE-ubuntu${ubuntu_version}.tar.gz && \
    rm swift-5.7.3-RELEASE-ubuntu${ubuntu_version}.tar.gz && \
    mv swift-5.7.3-RELEASE-ubuntu${ubuntu_version} /swift
ENV PATH=/swift/usr/bin:$PATH
```

File: .dockerignore
Comment (kou) [inline .dockerignore:66]: I think that this is needless because this is not used in Dockerfile. We can access `swift/Arrow` in `docker run` because we mount the source directory to `/arrow` in `docker run`.

File: docker-compose.yml
Comment (kou) [inline docker-compose.yml:886]: Do we need to use `ubuntu-cpp` as base image?

Should we use a Swift's Docker image instead? https://hub.docker.com/_/swift/?tab=tags

File: swift/Arrow/README.md
Comment (kou) [inline swift/Arrow/README.md:20]: How about adding "Swift" such as "Apache Arrow Swift"?

```suggestion
# Apache Arrow Swift
```

File: swift/Arrow/README.md
Comment (kou) [inline swift/Arrow/README.md:24]:
```suggestion
## Status
```

The following headings should have one more `#` too.

File: swift/Arrow/Package.swift
Comment (kou) [inline swift/Arrow/Package.swift:2]: Could you add license header?

File: swift/Arrow/Package.swift
Comment (kou) [inline swift/Arrow/Package.swift:6]: Can we specify package license?

File: .github/workflows/swift.yml
Comment (abandy) [reply .github/workflows/swift.yml:62]: Will remove matrix.ubuntu

File: swift/Arrow/README.md
Comment (abandy) [reply swift/Arrow/README.md:20]: Will do

File: swift/Arrow/Package.swift
Comment (abandy) [reply swift/Arrow/Package.swift:6]: I am not sure what you mean.  I didn't see an attribute on Package for a license.

File: swift/Arrow/Package.swift
Comment (abandy) [reply swift/Arrow/Package.swift:2]: Will do

File: .dockerignore
Comment (abandy) [reply .dockerignore:66]: Gocha, will remove.

File: ci/docker/linux-apt-swift.dockerfile
Comment (abandy) [reply ci/docker/linux-apt-swift.dockerfile:44]: Will do.

File: docker-compose.yml
Comment (abandy) [reply docker-compose.yml:886]: Good idea, will do.

File: .dockerignore
Comment (abandy) [reply .dockerignore:66]: I tried removing the above but it fails with the below error when I run 'archery docker run ubuntu-swift bash'

` > [2/3] ADD swift/ /arrow/swift:
------
failed to compute cache key: "/swift" not found: not found
ERROR: Service ubuntu-swift failed to build : Build failed
Error: docker-compose --file /Users/alvabandy/Documents/GitHub/arrow/docker-compose.yml build --build-arg BUILDKIT_INLINE_CACHE=1 ubuntu-swift exited with a non-zero exit code 1, see the process log above.`

Comment (kou) [reply .dockerignore:65]: We can remove `ADD swift/ /arrow/swift` because we mount `swift/` in `docker run`.

File: swift/Arrow/Package.swift
Comment (kou) [reply swift/Arrow/Package.swift:23]: Most package formats can include license information.
Examples:
* Python package:  https://pypi.org/project/pyarrow/ " License OSI Approved :: Apache Software License"
* Ruby package: https://rubygems.org/gems/red-arrow "License: Apache-2.0"

I'm sorry but I'm not familiar with Swift package format. If Swift package format supports license information, we should use it.

File: .dockerignore
Comment (abandy) [reply .dockerignore:65]: Ok, will remove.

File: swift/Arrow/Package.swift
Comment (abandy) [reply swift/Arrow/Package.swift:23]: I see what you mean.  I have looked at some other repos and the Package documentation and it doesn't look like the Package.swift file contains any licensing info besides the one included in the header.  

Once this PR has been merged, I will look at registering this package in https://swiftpackageregistry.com.

Comment (abandy) [reply swift/Arrow/Package.swift:23]: Gotcha, thanks for letting me know.

File: .dockerignore
Comment (abandy) [reply .dockerignore]: Makes sense, will do.

File: ci/docker/linux-apt-swift.dockerfile
Comment (abandy) [reply ci/docker/linux-apt-swift.dockerfile:44]: Makes sense :), will remove.

File: docker-compose.yml
Comment (abandy) [reply docker-compose.yml]: Will do.

File: .github/workflows/swift.yml
Comment (kou) [inline .github/workflows/swift.yml:64]:
```suggestion
      name: AMD 64 Ubuntu 20.04 Swift 5.7
      runs-on: ubuntu-latest
      if: ${{ !contains(github.event.pull_request.title, 'WIP') }}
      timeout-minutes: 15
```

File: .github/workflows/swift.yml
Comment (kou) [inline .github/workflows/swift.yml:76]:
```suggestion
            key: ubuntu-20.04-swift-${{ hashFiles('swift/**') }}
            restore-keys: ubuntu-20.04-swift-
```

File: .github/workflows/swift.yml
Comment (kou) [inline .github/workflows/swift.yml:86]:
```suggestion
            archery docker run ubuntu-swift
```

File: .github/workflows/swift.yml
Comment (kou) [inline .github/workflows/swift.yml:93]:
```suggestion
          run: archery docker push ubuntu-swift
```

File: docker-compose.yml
Comment (kou) [inline docker-compose.yml:144]:
```suggestion
  - ubuntu-swift
```

(We need to put this between `ubuntu-r-valgrind` and `ubuntu-verify-rc` to keep order.)

File: .github/workflows/swift.yml
Comment (abandy) [reply .github/workflows/swift.yml:64]: Will do.

File: .github/workflows/swift.yml
Comment (abandy) [reply .github/workflows/swift.yml:76]: Will do.

File: .github/workflows/swift.yml
Comment (abandy) [reply .github/workflows/swift.yml:86]: Will do.

File: .github/workflows/swift.yml
Comment (abandy) [reply .github/workflows/swift.yml:93]: Will do.

File: docker-compose.yml
Comment (abandy) [reply docker-compose.yml:144]: I see, I will move.

File: docker-compose.yml
Comment (raulcd) [inline docker-compose.yml:889]: I have no experience with swift but don't we have to build first? I can't see us using the added `swift_build.sh` script anywhere, is it necessary?

**Review by raulcd — COMMENTED** [review]
Thanks for the PR!

File: docker-compose.yml
Comment (abandy) [reply docker-compose.yml:889]: Good point, "swift test" will start the build.  I will remove the swift_build.sh script.

File: .github/workflows/swift.yml
Comment (assignUser) [inline .github/workflows/swift.yml:24]: this seems a bit broad to me, in the interest of being conservative with our usage of the common apache runners I think being more specific here would be good

File: .github/workflows/swift.yml
Comment (assignUser) [inline .github/workflows/swift.yml:45]:
```suggestion
```

File: .github/workflows/swift.yml
Comment (assignUser) [inline .github/workflows/swift.yml:62]: Is this actually needed? Does swift cache anything in one of the applicable volumes like the c++ builds do with ccache?

File: docker-compose.yml
Comment (assignUser) [inline docker-compose.yml:886]: Is this needed?

File: swift/Arrow/Package.swift
Comment (assignUser) [reply swift/Arrow/Package.swift:23]: @abandy I am not familar with swift how is it distributed? Are the binary artifacts that need to be built and distributed or is it more in line with go where a github tag is the only thing required?

**Review by assignUser — COMMENTED** [review]
Thanks for the PR I have added a few comments. I have also created the `Component: Swift` label and add it.

File: swift/Arrow/Package.swift
Comment (abandy) [reply swift/Arrow/Package.swift:23]: It would be more inline with go.  Packages are included using git url and tag.

File: .github/workflows/swift.yml
Comment (abandy) [reply .github/workflows/swift.yml:45]: Will remove.

File: .github/workflows/swift.yml
Comment (abandy) [reply .github/workflows/swift.yml:24]: Will change to - 'ci/docker/*swift*'

File: .github/workflows/swift.yml
Comment (abandy) [reply .github/workflows/swift.yml:62]: It shouldn't, I will remove.

**Comment by abandy** [comment]
Is it possible to get this approved soon (so I can start on adding the IPC reader/writer)?

File: docker-compose.yml
Comment (kou) [inline docker-compose.yml:883]:
```suggestion
      dockerfile: ci/docker/ubuntu-swift.dockerfile
```

**Comment by kou** [comment]
We can merge this after we remove needless Ubuntu versions.

File: .github/workflows/swift.yml
Comment (abandy) [reply .github/workflows/swift.yml]: Will do.

File: docker-compose.yml
Comment (abandy) [reply docker-compose.yml]: Will change.

File: docker-compose.yml
Comment (abandy) [reply docker-compose.yml]: Will change.

File: docker-compose.yml
Comment (kou) [inline docker-compose.yml:880]:
```suggestion
```

**Review by kou — COMMENTED** [review]
Sorry...
Two more things...

File: docker-compose.yml
Comment (abandy) [reply docker-compose.yml:880]: No worries, good catch! Will remove.

File: ci/docker/ubuntu-swift.dockerfile
Comment (kou) [inline ci/docker/ubuntu-swift.dockerfile:21]: I think that this `RUN` does nothing.
Can we remove this? Furthermore, can we use `swift:5.7.3` directly?

**Comment by kou** [comment]
Sorry. One my comment wasn't submitted. Could you confirm it?

File: ci/docker/ubuntu-swift.dockerfile
Comment (abandy) [reply ci/docker/ubuntu-swift.dockerfile]: I removed the RUN command.  I tried using swift:5.7.3 in the docker-compose file but it fails as it requires a url to the repo I believe.  Looks like the from in the swift.dockerfile is able to resolve the url without having to hard code it.

Comment (kou) [reply ci/docker/ubuntu-swift.dockerfile]: I've pushed a change to remove this file.

**Review by kou — APPROVED** [review]
+1

File: ci/docker/ubuntu-swift.dockerfile
Comment (abandy) [reply ci/docker/ubuntu-swift.dockerfile]: Nice, thank you!

**Pull request merged**
